plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -229,81 +228,79 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
         List<ConnectorSplit> splits = new ArrayList<>(maxSize);
         while (splits.size() < maxSize && (fileTasksIterator.hasNext() || fileScanIterator.hasNext())) {
             if (!fileTasksIterator.hasNext()) {
-                FileScanTask wholeFileTask = fileScanIterator.next();
-                if (wholeFileTask.deletes().isEmpty() && noDataColumnsProjected(wholeFileTask)) {
-                    fileTasksIterator = List.of(wholeFileTask).iterator();
+                if (limit.isPresent() && limit.getAsLong() <= outputRowsLowerBound) {
```

Comment:
why can't we have this test just after `saturatedAdd(outputRowsLowerBound, wholeFileTask.file().recordCount());`?

Reply:
At that point we have just started processing the file and haven't delivered the splits from it. Also, the splits are consumed in batch sizes given by the calling code. So we wait until processing the file is finished and apply the check before opening the next file.